art/estimators/speech_recognition/pytorch_icefall.py

```diff
@@ -0,0 +1,369 @@
+        if y is not None:
+            supervisions["text"] = [y[idx] for idx in indices]
+
+        feature_sorted = [feature_list[index] for index in indices]
+
+        feature = torch.zeros(
+            len(feature_sorted), feature_sorted[0].size(0), feature_sorted[0].size(1), device=self._device
+        )
+
+        for i in range(len(x)):
+            feature[i, :, : feature_sorted[i].size(1)] = feature_sorted[i]
+
+        return feature.transpose(1, 2), supervisions, indices
+
+    def to_training_mode(self) -> None:
+        """
+        Put the estimator in the training mode.
+        """
+        self.transducer_model.train()
+
+    @property
+    def sample_rate(self) -> int:
+        """
+        Get the sampling rate.
+
+        :return: The audio sampling rate.
+        """
+        return self._sampling_rate
+
+    @property
+    def input_shape(self) -> Tuple[int, ...]:
+        """
+        Return the shape of one input sample.
+
+        :return: Shape of one input sample.
+        """
+        return self._input_shape  # type: ignore
+
+    @property
+    def model(self):
+        """
+        Get current model.
+
+        :return: Current model.
+        """
+        return self._model
+
+    @property
+    def device(self) -> "torch.device":
+        """
+        Get current used device.
+
+        :return: Current used device.
+        """
+        return self._device
+
+    def get_activations(
+        self, x: np.ndarray, layer: Union[int, str], batch_size: int, framework: bool = False
+    ) -> np.ndarray:
+        raise NotImplementedError
+
+    def compute_loss(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
+        raise NotImplementedError
+
+    def compute_loss_and_decoded_output(
+        self, masked_adv_input: "torch.Tensor", original_output: np.ndarray, **kwargs
+    ) -> Tuple["torch.Tensor", np.ndarray]:
+        """
+        Compute loss function and decoded output.
+
+        :param masked_adv_input: The perturbed inputs.
+        :param original_output: Target values of shape (nb_samples). Each sample in `original_output` is a string and
+                                it may possess different lengths. A possible example of `original_output` could be:
+                                `original_output = np.array(['SIXTY ONE', 'HELLO'])`.
+        :return: The loss and the decoded output.
+        """
+        from transducer.beam_search import greedy_search
+        import k2
+
+        assert len(original_output[0]) == 1
+        num_batch = len(original_output[0])
+        decoded_output = []
+
+        for sample_index in range(num_batch):
+            features, _, _ = self.transform_model_input(x=masked_adv_input[sample_index])
+            x_lens = torch.tensor([features.shape[1]]).to(torch.int32).to(self.device)
+            y = k2.RaggedTensor(original_output[sample_index])
+            loss = self.transducer_model(x=features, x_lens=x_lens, y=y)
```

Comment:
Should the loss not also be returned?

Reply:
I initially assumed so too but apprently not (if the other estimator implementations are anything to go by)